File: deno/app/fizz-buzz.test.ts
import { series, step } from "./fizz-buzz.ts";
import { assertEquals } from "./deps.dev.ts";
import { FizzBuzzValue } from "./fizz-buzz.d.ts";

Deno.test("11 is 11", () => {
  assertEquals(11, step(11));
});

Deno.test("12 is Fizz", () => {
  assertEquals("Fizz" as FizzBuzzValue, step(12));
});

Deno.test("25 is FizzBuzz", () => {
  assertEquals("Buzz" as FizzBuzzValue, step(25));
});

Deno.test("30 is FizzBuzz", () => {
  assertEquals("FizzBuzz" as FizzBuzzValue, step(30));
});

Deno.test("from -10 to 10 are 21 items", () => {
  const values = series(-10, 10);

  assertEquals(values.length, 21);
});

Deno.test("from -10 to 0 first item is -10", () => {
  const values = series(-10, 0);

  assertEquals(values[0].step, -10);
});

Deno.test("from 0 to 10 last item is 10", () => {
  const values = series(0, 10);

  assertEquals(11, values.length);
  assertEquals(values.at(-1)?.step, 10);
});

File: go/internal/core/fizz_buzz_test.go
package core_test

import (
	"fizz-buzz/internal/core"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestStep(t *testing.T) {
	type Test_Case struct {
		input  int
		output string
	}

	cases := []Test_Case{
		{11, "11"},
		{12, "Fizz"},
		{25, "Buzz"},
		{30, "FizzBuzz"},
	}

	for _, test_case := range cases {
		assert.Equal(t, test_case.output, core.Step(test_case.input))
	}
}

func TestSeries(t *testing.T) {
	var values []core.StepType

	//  total elements
	values = core.Series(-10, 10)
	assert.Len(t, values, 21)

	// first item (even negative)
	values = core.Series(-10, 0)
	assert.Equal(t, -10, values[0].Step)

	// last item
	values = core.Series(0, 10)
	assert.Equal(t, 10, values[len(values)-1].Step)
}
